python/gps/algorithm/cost/ioc_layers.py

```diff
@@ -50,9 +55,78 @@ def forward(self, bottom, top):
 
         # TODO(kevin) implement this (look at ioc_loss_layer.cpp in caffe)
         # python/gps/algorithm/policy_opt/policy_layers.py might also be helpful
+        loss = 0.0
+        dc = self.demo_counts_
+        sc = self.sample_counts_
+
+        # log importance weights of demos and samples.
+        d_log_iw = bottom[2].data
+        s_log_iw = bottom[3].data
+
+        # Sum over time and compute max value for safe logsum.
+        for i in xrange(self.num_demos):
+            dc[i] = 0.0
+            for t in xrange(self.T):
+                dc[i] += 0.5 * bottom[0].data[i, t, 0] # Not sure why the feature should be 0
+            loss += dc[i]
+            # Add importance weight to demo feature count. Will be negated.
+            dc[i] += d_log_iw[i]
+        # Divide by number of demos.
+        loss /= self.num_demos
+
+        max_val = -dc[0]
+        for i in xrange(self.num_samples):
+            sc[i] = 0.0
+            for t in xrange(self.T):
+                sc[i] += 0.5 * bottom[1].data[i, t, 0]
+            # Add importance weight to demo feature count. Will be negated.
+            sc[i] += s_log_iw[i]
+            if -sc[i] > max_val:
+                max_val = -sc[i]
+
+        # Do a safe log-sum-exp operation.
+        append_demos = self.layer_params['append_demos'] # We need to set this somewhere?
+        if append_demos:
+            for i in xrange(self.num_demos):
+                if -dc[i] > max_val:
+                    max_val = -dc[i]
+            for i in xrange(self.num_demos):
+                dc[i] = -dc[i] - max_val
+        for i in xrange(self.num_samples):
+            sc[i] = -sc[i] - max_val
+        dc = np.exp(dc)
+        sc = np.exp(sc)
+
+        self.partition = 0.0
+        if append_demos:
+            self.partition += np.sum(dc, axis = 0)
+        self.partition += np.sum(sc, axis = 0)
+        loss += np.log(self.partition) + max_val
+        top[0].data[...] = loss
+        # Need to increment iteration here?
```

Comment:
No, the iteration_ variable was for some debugging code, where I as printing out the effective sample size (ess) every N iterations. Not needed anymore.